Rewrite Error.p.stack (#30908)
mdn/content: files/en-us/web/javascript/reference/global_objects/error/stack/index.md

{{JSRef}} {{Non-standard_Header}}

> > **Note:** The `stack` property is de-facto implemented by most, if not all, JavaScript engines. [The standard committee is looking to standardize this feature too](https://github.com/tc39/proposal-error-stacks). You cannot rely on the precise content of the stack string due to implementation inconsistencies, but you can usually assume it exists and use it for debugging purposes.

Reviewer: ```suggestion
> **Note:** The `stack` property is de facto implemented by all major JavaScript engines, and [the JavaScript standards committee is looking to standardize it](https://github.com/tc39/proposal-error-stacks). You cannot rely on the precise content of the stack string due to implementation inconsistencies, but you can generally assume it exists and use it for debugging purposes.
```

```js
function foo() {
  bar();
}

function bar() {
  baz();
}

function baz() {
  console.log(new Error().stack);
}

foo();
```

```plain
#### JavaScriptCore
baz@filename.js:10:24
bar@filename.js:6:6
foo@filename.js:2:6
global code@filename.js:13:4

#### SpiderMonkey
baz@filename.js:10:15
bar@filename.js:6:3
foo@filename.js:2:3
@filename.js:13:1

#### V8
Error
    at baz (filename.js:10:15)
    at bar (filename.js:6:3)
    at foo (filename.js:2:3)
    at filename.js:13:1
```

Different engines set this value at different times. Most modern engines set it when the {{jsxref("Error")}} object is created. This means you can get the full call stack information within a function using the following:

```js
function foo() {
  console.log(new Error().stack);
}
```

Without having to throw an error and then catch it.

In V8, the non-standard `Error.captureStackTrace()`, `Error.stackTraceLimit`, and `Error.prepareStackTrace()` APIs can be used to customize the stack trace. Read the [Stack trace API](https://v8.dev/docs/stack-trace-api) in the V8 docs for more information.

> Stack frames can be things other than explicit function calls, too. For example, event listeners, timeout jobs, and promise handlers all begin their own call chain. Source code within {{jsxref("Global_Objects/eval", "eval()")}} and {{jsxref("Function")}} constructor also appear in the stack:

Author: It's quite hard to describe, or maybe I don't understand this deeply enough. I am writing a page for execution contexts (maybe in another year? 😛) https://github.com/mdn/content/issues/5006 and I'll talk more about stacks then.

> Stack frames can be things other than explicit function calls, too. For example, event listeners, timeout jobs, and promise handlers all begin their own call chain. Source code within {{jsxref("Global_Objects/eval", "eval()")}} and {{jsxref("Function")}} constructor also appear in the stack:

Reviewer: Is this any use: https://stackoverflow.com/questions/10057443/explain-the-concept-of-a-stack-frame-in-a-nutshell#10057535 ?

I wasn't thinking that it needed to be much, but it seems a little odd that there is no definition for terms such as these. I guess when you come to write your dedicated execution contexts page, you can link it to the definitions in there. I'll let this pass for now, as the definition of it is implied enough to not seriously impair understanding.

```js
console.log(new Function("return new Error('Function failled')")().stack);
console.log("====");
console.log(eval("new Error('eval failed')").stack);
```

```plain
#### JavaScriptCore
anonymous@
global code@filename.js:1:65
====
eval code@
eval@[native code]
global code@filename.js:3:17

#### SpiderMonkey
anonymous@filename.js line 1 > Function:1:8
@filename.js:1:65

====
@filename.js line 3 > eval:1:1
@filename.js:3:13

#### V8

> Error: Function failled

Reviewer: ```suggestion
Error: Function failed
```

    at eval (eval at <anonymous> (filename.js:1:13), <anonymous>:1:8)
    at filename.js:1:65
====
Error: eval failed
    at eval (eval at <anonymous> (filename.js:3:13), <anonymous>:1:1)
    at filename.js:3:13
```

> You can also use the `//# sourceURL` directive to name an eval source. For Firefox, see also the [Debug eval sources](https://firefox-source-docs.mozilla.org/devtools-user/debugger/how_to/debug_eval_sources/index.html) docs and the [Naming `eval` Scripts with the `//# sourceURL` Directive](https://fitzgeraldnick.com/2014/12/05/name-eval-scripts.html) blog post.

Author: I don't know but I think it is Firefox-specific. After all, devtools are always implementation-inconsistent. I just copied this from the paragraph that was in examples.

> You can also use the `//# sourceURL` directive to name an eval source. For Firefox, see also the [Debug eval sources](https://firefox-source-docs.mozilla.org/devtools-user/debugger/how_to/debug_eval_sources/index.html) docs and the [Naming `eval` Scripts with the `//# sourceURL` Directive](https://fitzgeraldnick.com/2014/12/05/name-eval-scripts.html) blog post.

Reviewer: Oh, OK. I've edited my suggested change to suit.

> You can also use the `//# sourceURL` directive to name an eval source. For Firefox, see also the [Debug eval sources](https://firefox-source-docs.mozilla.org/devtools-user/debugger/how_to/debug_eval_sources/index.html) docs and the [Naming `eval` Scripts with the `//# sourceURL` Directive](https://fitzgeraldnick.com/2014/12/05/name-eval-scripts.html) blog post.

Reviewer: Committed.

## Examples

### Using the stack property

> The following script demonstrates the use of `stack` property. It runs in your browser, so you can check how the stack trace looks in your browser.

Author: The key here is that the stack trace is "from your browser", so if you don't use one of the browsers described above, or if you use a different version, you can verify what its stack structure looks like.

> The following script demonstrates the use of `stack` property. It runs in your browser, so you can check how the stack trace looks in your browser.

Reviewer: Edit suggestion updated to suit. I'm not quite so sure about this one, so I'll let you check this before committing.
